river/command/output.zig

```diff
@@ -0,0 +1,105 @@
+
+pub fn focusOutput(
+    allocator: *std.mem.Allocator,
+    seat: *Seat,
+    args: []const []const u8,
+    out: *?[]const u8,
+) Error!void {
+    if (args.len < 2) return Error.NotEnoughArguments;
+    if (args.len > 2) return Error.TooManyArguments;
+
+    // If the noop output is focused, there are no other outputs to switch to
+    if (seat.focused_output == &server.root.noop_output) {
+        std.debug.assert(server.root.outputs.len == 0);
+        return;
+    }
+
+    seat.focusOutput((try getOutput(seat, args[1])) orelse return);
+    seat.focus(null);
+    server.root.startTransaction();
+}
+
+pub fn sendToOutput(
+    allocator: *std.mem.Allocator,
+    seat: *Seat,
+    args: []const []const u8,
+    out: *?[]const u8,
+) Error!void {
+    if (args.len < 2) return Error.NotEnoughArguments;
+    if (args.len > 2) return Error.TooManyArguments;
+
+    // If the noop output is focused, there is nowhere to send the view
+    if (seat.focused_output == &server.root.noop_output) {
+        std.debug.assert(server.root.outputs.len == 0);
+        return;
+    }
+
+    if (seat.focused == .view) {
+        const destination_output = (try getOutput(seat, args[1])) orelse return;
+        seat.focused.view.sendToOutput(destination_output);
+
+        // Handle the change and focus whatever's next in the focus stack
+        seat.focus(null);
+        seat.focused_output.arrangeViews();
+        destination_output.arrangeViews();
+        server.root.startTransaction();
+    }
+}
+
+/// Find an output adjacent to the currently focused based on either logical or
+/// spacial direction
+fn getOutput(seat: *Seat, str: []const u8) !?*Output {
+    if (std.meta.stringToEnum(Direction, str)) |direction| { // Logical direction
+        // Focus the next/prev output in the list if there is one, else wrap
+        const focused_node = @fieldParentPtr(std.TailQueue(Output).Node, "data", seat.focused_output);
+        return switch (direction) {
+            .next => if (focused_node.next) |node| &node.data else &server.root.outputs.first.?.data,
+            .previous => if (focused_node.prev) |node| &node.data else &server.root.outputs.last.?.data,
+        };
+    } else if (std.meta.stringToEnum(wl.OutputLayout.Direction, str)) |direction| { // Spacial direction
+        const focus_box = server.root.output_layout.getBox(seat.focused_output.wlr_output) orelse return null;
+        const wlr_output = server.root.output_layout.adjacentOutput(
+            direction,
+            seat.focused_output.wlr_output,
+            @intToFloat(f64, focus_box.x + @divFloor(focus_box.width, 2)),
+            @intToFloat(f64, focus_box.y + @divFloor(focus_box.height, 2)),
+        ) orelse return null;
+        var it = server.root.outputs.first;
+        while (it) |node| : (it = node.next) {
+            if (node.data.wlr_output == wlr_output) {
+                return &node.data;
+            }
+        }
+        return null;
```

Comment:
No need to iterate here, just `@ptrToInt(*Output, wlr_output.data);`

Reply:
I always forget the wlroots objects have user data.

Reply:
But you do mean `@intToPtr()`, right?

Reply:
> But you do mean @intToPtr(), right?

Yeah definitely